### PR TITLE
Add new api route for referral flow

### PR DESCRIPTION
**Pull request by kewitham**
[Issue 1577](https://github.com/usdoj-crt/crt-portal-management/issues/1577)

## What does this change?
This PR adds an API route for the new referral flow so users can send, print, or copy response templates without refreshing the page.

### Author

+ [ ] If this is a story, run locally and check to make sure all Acceptance Criteria are met. (If any criteria are unclear, ask about them.).
+ [ ] Check for, document, and establish a testing plan for any behavior that may vary across environments or is otherwise difficult to test.
+ [ ] Check for [accessibility](https://github.com/usdoj-crt/crt-portal/blob/develop/docs/a11y_plan.md).
+ [ ] [Tests pass](https://github.com/USDOJ/crt-portal/#tests).

### Reviewer

+ [ ] If this is a story, run locally and check to make sure all Acceptance Criteria are met. (If any criteria are unclear, ask about them.).
+ [ ] Check for any behavior that may vary across environments or is difficult to test, and ensure that it is well-understood, documented, and that there is a testing plan in place.
+ [ ] Re-check for [accessibility](https://github.com/usdoj-crt/crt-portal/blob/develop/docs/a11y_plan.md).
+ [ ] [Tests pass](https://github.com/USDOJ/crt-portal/#tests).

## Notes for reviewer:

See [PR instructions doc](https://github.com/usdoj/crt-portal/blob/master/docs/pull_requests.md) for full pull request review instructions.

### CODE REVIEW

File: crt_portal/api/views.py [hunk 1]
```diff
@@ -284,3 +285,32 @@ def get(self, request):
...
+        preview = {}
+        if action == 'send':
+            try:
+                email_response = crt_send_mail(report, template)
+                if email_response:
+                    response = f'Sent referral email template #{template.id} to report #{report.id}'
+                else:
+                    response = f'Referral email template #{template.id} failed to send to report #{report.id}'
+            except Exception as e:
+                response = f'Referral email template #{template.id} failed to send to report #{report.id}: {e}'
```
Comment (okkays) [inline crt_portal/api/views.py:315]: this try-except is a bit broad (nothing is ever going to go wrong with the `if email_response else` portion unless there's a syntax error in the code, in which case you _don't_ want to catch the exception.

These kinds of nested blocks can be simplified a bit if you're able to return early. For example, maybe instead:

```py
if action == 'send':
  response = {'preview': preview}
  try:
    email_response = crt_send_mail(report, template)
  except Exception as e:
    return Response({'email_response': f'Referral email template #{template.id} failed to send to report #{report.id}: {e}', **response})
  if not email_response:
    return Response({'email_response': f'Referral email template #{template.id} failed to send to report #{report.id}', **response})
  return Response({'email_response': f'Sent referral email template #{template.id} to report #{report.id}', **response})
```

You could further reduce nesting (especially if 'copy' and 'print' turn out to be complicated) by making a helper function for the `if action == 'send'` portion

**Review by okkays — CHANGES_REQUESTED** [review]

File: crt_portal/api/views.py [hunk 2]
```diff
@@ -284,3 +285,32 @@ def get(self, request):
...
+        template = get_object_or_404(ResponseTemplate, pk=template_id)
+        action = request.data['action']
+        preview = {}
+        if action == 'send':
+            try:
+                email_response = crt_send_mail(report, template)
+                if email_response:
+                    response = f'Sent referral email template #{template.id} to report #{report.id}'
+                else:
+                    response = f'Referral email template #{template.id} failed to send to report #{report.id}'
```
[the remark this answers is not in the document]
Comment (ilodidoj) [reply crt_portal/api/views.py:313]: @kewitham @okkays we do get reasons for failure some times, it's the error field under TMS Emails > Error message field. But I think this happen later once we pass the msg to TMS client.

[the remark this answers is not in the document]
Comment (kewitham) [reply crt_portal/api/views.py:313]: crt_send_mail function calls the [send_mail function](https://docs.djangoproject.com/en/4.2/topics/email/) which only returns 0 or 1 depending on how many messages are successful so I'm not sure how we would include the additional information.

[the remark this answers is not in the document]
Comment (okkays) [reply crt_portal/api/views.py:313]: Sounds good. Then I'd just return a `502` if something goes wrong (indicating the mail service failed to send)

File: crt_portal/api/views.py [hunk 3]
```diff
@@ -284,3 +285,36 @@ def get(self, request):
...
+
+    def post(self, request):
+        report_id = request.data['report_id']
+        report = get_object_or_404(Report, pk=report_id)
+        template_id = request.data['template_id']
+        template = get_object_or_404(ResponseTemplate, pk=template_id)
+        action = request.data['action']
+        preview = {'preview': {}}
+        if not action:
+            return Response({'response': '500: No action provided', **preview})
```
Comment (okkays) [inline crt_portal/api/views.py:308]: I don't think this actually returns a 500

also - 500 should be in the event of a unexpected server error (our code broke) this should probably be 400

https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_client_errors

https://docs.djangoproject.com/en/4.2/ref/request-response/#jsonresponse-objects

```suggestion
            return JsonResponse(
              {
                'message': 'No action provided',
                **preview,
              }, status=400
              )
```

**Review by okkays — CHANGES_REQUESTED** [review]

File: crt_portal/api/views.py [hunk 4]
```diff
@@ -284,3 +285,41 @@ def get(self, request):
...
+              {
+                'message': 'No action provided',
+                **preview,
+              }, status=400
+              )
+        if action == 'send':
+            try:
+                email_response = crt_send_mail(report, template)
+            except Exception as e:
+                return Response({'response': f'502: Referral email template #{template.id} failed to send to report #{report.id}: {e}', **preview})
```
Comment (okkays) [inline crt_portal/api/views.py:318]: Sorry, here as well

File: crt_portal/api/views.py [hunk 5]
```diff
@@ -284,3 +285,41 @@ def get(self, request):
...
+            if not email_response:
+                return Response({'response': f'502: Referral email template #{template.id} failed to send to report #{report.id}', **preview})
```
Comment (okkays) [inline crt_portal/api/views.py:320]: - And here

**Review by okkays — CHANGES_REQUESTED** [review]
sorry for the extra back and forth, lgtm after this

**Review by okkays — APPROVED** [review]

**Comment by okkays** [comment]
(Approved pending test updates to match the new statuses!)

**Comment by ilodidoj** [comment]
@kewitham @okkays Probably need to fix the unit test issue before merging, CI/CD failure needs to be resolved.

**Pull request merged**